src/main/java/com/datatorrent/netlet/WriteOnlyClient.java

```diff
@@ -0,0 +1,239 @@
+    public Slice poll()
+    {
+      return queue.poll();
+    }
+
+    @Override
+    public Slice peek()
+    {
+      return queue.peek();
+    }
+  }
+
+  protected final ByteBuffer writeBuffer;
+  protected Queue<Slice> sendQueue;
+  protected final SpscArrayQueue<Slice> freeQueue;
+  protected boolean isWriteEnabled = true;
+  protected Lock lock = new ReentrantLock();
+
+  public WriteOnlyClient()
+  {
+    this(64 * 1024, 1024);
+  }
+
+  public WriteOnlyClient(final int writeBufferCapacity, final int sendQueueCapacity)
+  {
+    this(ByteBuffer.allocateDirect(writeBufferCapacity), sendQueueCapacity);
+  }
+
+  public WriteOnlyClient(final ByteBuffer writeBuffer, final int sendQueueCapacity)
+  {
+    this.writeBuffer = writeBuffer;
+    sendQueue = new SpscArrayQueue<Slice>(sendQueueCapacity);
+    freeQueue = new SpscArrayQueue<Slice>(sendQueueCapacity);
+  }
+
+  @Override
+  public void connected()
+  {
+    super.connected();
+    shutdownIO(true);
+    suspendReadIfResumed();
+    try {
+      lock.lock();
+      isWriteEnabled = !isWriteSuspended();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public void unregistered(SelectionKey key)
+  {
+    try {
+      lock.lock();
+      if (sendQueue.peek() == null && !isWriteEnabled) {
+        super.unregistered(key);
+      } else {
+        sendQueue = new ForwardingQueue(sendQueue)
+        {
+          @Override
+          public boolean offer(Slice f)
+          {
+            throw new RuntimeException("Client " + this + " does not accept new data.");
+          }
+        };
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public final void read() throws IOException
+  {
+    if (suspendReadIfResumed()) {
+      logger.warn("{} OP_READ should be disabled", this);
+    } else {
+      logger.error("{} read is not expected", this);
+    }
+  }
+
+  @Override
+  public void write() throws IOException
+  {
+    /*
+     * at first when we enter this function, our buffer is in fill mode.
+     */
+    int remaining = writeBuffer.remaining();
+    Slice slice;
+    while ((slice = sendQueue.peek()) != null) {
+      while (remaining < slice.length) {
+        if (remaining > 0) {
+          writeBuffer.put(slice.buffer, slice.offset, remaining);
+          slice.offset += remaining;
+          slice.length -= remaining;
+        }
+        if (channelWrite() == 0) {
+          return;
+        }
+        remaining = writeBuffer.remaining();
+      }
+      writeBuffer.put(slice.buffer, slice.offset, slice.length);
+      slice.buffer = null;
+      remaining -= slice.length;
+      freeQueue.offer(sendQueue.poll());
+    }
+    channelWrite();
+  }
+
+  protected int channelWrite() throws IOException
+  {
+    if (writeBuffer.flip().remaining() > 0) {
+      final SocketChannel channel = (SocketChannel)key.channel();
+      final int write = channel.write(writeBuffer);
+      writeBuffer.compact();
+      return write;
+    } else {
```

Comment:
Why subject to the vagaries of delay from the nio API for suspend and resume when there is no data to be sent. Why not leave write enabled? Even when write is called your simple boolean checks in write() and channelWrite() will skip quickly without much overhead.